### PR TITLE
CHE-1218: Add TCK tests for User/Profile

### PR DESCRIPTION
**Pull request by voievodin**
The approach is a bit different from the `GitConnection` TCK tests. It is based on the _Guice_ modules which should be provided by the TCK clients. 

@skabashnyuk, @garagatyi, @sleshchenko please review

Merge of this PR depends on #1393

### CODE REVIEW

File: core/commons/che-core-commons-test/pom.xml [hunk 1]
```diff
@@ -21,10 +21,30 @@
     <artifactId>che-core-commons-test</artifactId>
     <packaging>jar</packaging>
     <name>Che Core :: Commons :: Utilities for tests</name>
+    <properties>
+        <findbugs.failonerror>true</findbugs.failonerror>
```
Comment (garagatyi) [inline core/commons/che-core-commons-test/pom.xml:25]: Awesome!

File: core/commons/che-core-commons-test/pom.xml [hunk 2]
```diff
@@ -21,10 +21,30 @@
...
+    </properties>
     <dependencies>
         <dependency>
+            <groupId>com.google.inject</groupId>
+            <artifactId>guice</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>javax.inject</groupId>
+            <artifactId>javax.inject</artifactId>
```
Comment (benoitf) [inline core/commons/che-core-commons-test/pom.xml]: should it be provided as well ? (like guava)

File: core/commons/che-core-commons-test/pom.xml [hunk 3]
```diff
@@ -21,10 +21,30 @@
     <artifactId>che-core-commons-test</artifactId>
     <packaging>jar</packaging>
     <name>Che Core :: Commons :: Utilities for tests</name>
+    <properties>
+        <findbugs.failonerror>true</findbugs.failonerror>
+    </properties>
     <dependencies>
         <dependency>
+            <groupId>com.google.inject</groupId>
```
Comment (benoitf) [inline core/commons/che-core-commons-test/pom.xml]: should it be provided as well ? (like guava)

File: core/commons/che-core-commons-test/src/main/java/org/eclipse/che/commons/test/tck/TckModule.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+public abstract class TckModule extends AbstractModule {
+
+    /**
+     * It is guaranteed that this field is always present and
+     * can be reused by implementation, the value is equal to the
+     * {@link IModuleFactory#createModule(ITestContext, Class)} first
+     * parameter and will be set by {@link TckModuleFactory} immediately after module
+     * implementation is loaded by {@link ServiceLoader}.
+     */
+    protected ITestContext testContext;
```
Comment (benoitf) [inline core/commons/che-core-commons-test/src/main/java/org/eclipse/che/commons/test/tck/TckModule.java]: do we need protected field or a protected getter ?

File: core/commons/che-core-commons-test/src/main/java/org/eclipse/che/commons/test/tck/TckModuleFactory.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+ * @see org.testng.annotations.Guice
+ * @see IModuleFactory
+ */
+public class TckModuleFactory implements IModuleFactory {
+
+    @Override
+    public Module createModule(ITestContext context, Class<?> testClass) {
+        final Iterator<TckModule> moduleIterator = ServiceLoader.load(TckModule.class).iterator();
+        if (!moduleIterator.hasNext()) {
+            throw new IllegalStateException(format("Couldn't find a TckModule configuration. " +
```
Comment (benoitf) [inline core/commons/che-core-commons-test/src/main/java/org/eclipse/che/commons/test/tck/TckModuleFactory.java:99]: why do we have string concatenation in format ?

File: core/commons/che-core-commons-test/src/test/resources/META-INF/services/org.eclipse.che.commons.test.tck.TckModule [hunk 6]
```diff
@@ -0,0 +1,2 @@
+org.eclipse.che.commons.test.tck.TestModule1
+org.eclipse.che.commons.test.tck.TestModule2
```
Comment (benoitf) [inline core/commons/che-core-commons-test/src/test/resources/META-INF/services/org.eclipse.che.commons.test.tck.TckModule:2]: if we can add empty line at eof ?

File: core/commons/che-core-commons-test/src/test/resources/META-INF/services/org.testng.ITestNGListener [hunk 7]
```diff
@@ -0,0 +1 @@
+org.eclipse.che.commons.test.tck.DBServerListener
```
Comment (benoitf) [inline core/commons/che-core-commons-test/src/test/resources/META-INF/services/org.testng.ITestNGListener:1]: if we can add empty line at eof ?

File: core/commons/che-core-commons-test/src/main/java/org/eclipse/che/commons/test/tck/TckModuleFactory.java (hunk 5 above)
Comment (voievodin) [inline core/commons/che-core-commons-test/src/main/java/org/eclipse/che/commons/test/tck/TckModuleFactory.java:99]:
```
"Couldn't find a TckModule configuration. " +
"You probably forgot to configure resources/META-INF/services/%s, or even " +
"provide an implementation of the TckModule which is required by the tck test class %s"
```

The message is too long for a single line, this component is used only in tests so it won't affect application performance.
Your suggestion ?

Comment (benoitf) [inline core/commons/che-core-commons-test/src/main/java/org/eclipse/che/commons/test/tck/TckModuleFactory.java:99]: only test so ok, but else it might be put on a file

File: wsmaster/wsmaster-local/src/test/resources/META-INF/services/org.eclipse.che.commons.test.tck.TckModule [hunk 8]
```diff
@@ -0,0 +1 @@
+org.eclipse.che.api.local.LocalTckModule
```
Comment (benoitf) [inline wsmaster/wsmaster-local/src/test/resources/META-INF/services/org.eclipse.che.commons.test.tck.TckModule:1]: empty line if possible ?

File: core/commons/che-core-commons-test/pom.xml (hunk 2 above)
Comment (voievodin) [inline core/commons/che-core-commons-test/pom.xml]: no

File: core/commons/che-core-commons-test/pom.xml (hunk 3 above)
Comment (voievodin) [inline core/commons/che-core-commons-test/pom.xml]: no

**Comment by garagatyi** [comment]
Please consider moving classes related to repository testing to dedicated package `tck/repository`

**Comment by garagatyi** [comment]
LGTM. Framework looks really interesting. Good job!

File: core/commons/che-core-commons-test/pom.xml (hunk 2 above)
Comment (benoitf) [inline core/commons/che-core-commons-test/pom.xml]: Why then guava has provided and not this one ? API is one of the reason we're using provided

File: wsmaster/wsmaster-local/pom.xml [hunk 9]
```diff
@@ -49,11 +53,20 @@
             <artifactId>javax.inject</artifactId>
         </dependency>
         <dependency>
+            <groupId>javax.ws.rs</groupId>
+            <artifactId>javax.ws.rs-api</artifactId>
+            <version>2.0</version>
```
Comment (skabashnyuk) [inline wsmaster/wsmaster-local/pom.xml]: var

File: wsmaster/wsmaster-local/pom.xml [hunk 10]
```diff
@@ -114,6 +142,12 @@
             <scope>test</scope>
         </dependency>
         <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <version>1.9.5</version>
```
Comment (skabashnyuk) [inline wsmaster/wsmaster-local/pom.xml]: var

File: core/commons/che-core-commons-test/pom.xml (hunk 2 above)
Comment (voievodin) [inline core/commons/che-core-commons-test/pom.xml]: Can you please explain why this dependency should be provided?
Guava is provided because dependency analyzer will fail with an error 

```
[WARNING] Unused declared dependencies found:
[WARNING]    com.google.guava:guava:jar:18.0:compile
```

But we can't remove the dependency, beucase we will get the following error:

```
[ERROR] java.lang.NoClassDefFoundError: com/google/common/collect/Sets
[ERROR] at com.google.inject.internal.Errors.<clinit>(Errors.java:82)
[ERROR] at com.google.inject.internal.InternalInjectorCreator.<init>(InternalInjectorCreator.java:62)
[ERROR] at com.google.inject.Guice.createInjector(Guice.java:96)
[ERROR] at com.google.inject.Guice.createInjector(Guice.java:73)
[ERROR] at com.google.inject.Guice.createInjector(Guice.java:62)
[ERROR] at org.testng.internal.ClassImpl.getParentInjector(ClassImpl.java:159)
[ERROR] at org.testng.internal.ClassImpl.getInstanceFromGuice(ClassImpl.java:132)
[ERROR] at org.testng.internal.ClassImpl.getDefaultInstance(ClassImpl.java:110)
[ERROR] at org.testng.internal.ClassImpl.getInstances(ClassImpl.java:200)
[ERROR] at org.testng.internal.TestNGClassFinder.<init>(TestNGClassFinder.java:120)
[ERROR] at org.testng.TestRunner.initMethods(TestRunner.java:409)
[ERROR] at org.testng.TestRunner.init(TestRunner.java:235)
[ERROR] at org.testng.TestRunner.init(TestRunner.java:205)
[ERROR] at org.testng.TestRunner.<init>(TestRunner.java:153)
[ERROR] at org.testng.SuiteRunner$DefaultTestRunnerFactory.newTestRunner(SuiteRunner.java:536)
[ERROR] at org.testng.SuiteRunner.init(SuiteRunner.java:159)
[ERROR] at org.testng.SuiteRunner.<init>(SuiteRunner.java:113)
[ERROR] at org.testng.TestNG.createSuiteRunner(TestNG.java:1299)
[ERROR] at org.testng.TestNG.createSuiteRunners(TestNG.java:1286)
[ERROR] at org.testng.TestNG.runSuitesLocally(TestNG.java:1140)
[ERROR] at org.testng.TestNG.run(TestNG.java:1057)
[ERROR] at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:115)
[ERROR] at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeMulti(TestNGDirectoryTestSuite.java:205)
[ERROR] at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.execute(TestNGDirectoryTestSuite.java:108)
[ERROR] at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:111)
[ERROR] at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:203)
[ERROR] at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:155)
[ERROR] at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)

```

Comment (benoitf) [inline core/commons/che-core-commons-test/pom.xml]: I said that API should have scope provided not to remove guava

File: wsmaster/wsmaster-local/pom.xml (hunk 9 above)
Comment (sleshchenko) [inline wsmaster/wsmaster-local/pom.xml]: remove version

File: core/commons/che-core-commons-test/pom.xml (hunk 2 above)
Comment (voievodin) [inline core/commons/che-core-commons-test/pom.xml]: Yes i understand, i tried to explain why the guava scope is _provided_, can you please explain why api should use dependencies with provided scope, afaik we don't use such approach

Comment (benoitf) [inline core/commons/che-core-commons-test/pom.xml]: yes
https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#Dependency_Scope
provided
_This is much like compile, but indicates you expect the JDK or a container to provide the dependency at runtime. For example, when building a web application for the Java Enterprise Edition, you would set the dependency on the Servlet API and related Java EE APIs to scope provided because the web container provides those classes. This scope is only available on the compilation and test classpath, and is not transitive._

so to sum up, in Eclipse Che javax.inject, guice are provided by Eclipse Che framework as it's "core stuff" to every server component (the platform) so all dependencies should use provided on these "platform depedencies"

File: wsmaster/wsmaster-local/pom.xml (hunk 10 above)
Comment (benoitf) [inline wsmaster/wsmaster-local/pom.xml]: I would remove version

**Comment by voievodin** [comment]
Note that this branch is merged into CHE-1078 #1393
